### PR TITLE
Implements `findUserByMe` endpoint + trimmings

### PR DESCRIPTION
**Pull request by gbdubs**

### CODE REVIEW

File: cmd/server/pactasrv/conv_pacta_to_oapi.go
Comment (bcspragu) [inline cmd/server/pactasrv/conv_pacta_to_oapi.go:31]: I hear your point about keeping all the `conv` stuff in the `pactasrv` package so you can access these, but my counterpoint would be that determining the 'type' of an error truly is a concern of the server, and this should just return a `fmt.Errorf(...)` or some `conv`-specific error that the `srv` layer then converts into an `errorWhatever`

Functionally, it probably doesn't matter, just cleaner abstractions and easier testing

Comment (bcspragu) [inline cmd/server/pactasrv/conv_pacta_to_oapi.go:31]: I remember you had a reason for using `*api.Whatever` and `dereference` instead of just using structs directly, but it seems like a lot of contorting to be using pointers, but having to check for errors/disallow them, only to convert them back into not-pointers later

File: cmd/server/pactasrv/user.go
Comment (bcspragu) [inline cmd/server/pactasrv/user.go:92]: I would not use a `GetOrCreate` endpoint for something like this. `GetOrCreate`-ing a user should happen on a dedicated `login` or similar endpoint, it should be an error if a user tries to do anything else without an account in the DB.

There's a bunch of reasons to do this, but the main one is to have a single location where we create accounts in the DB.

File: db/sqldb/user.go
Comment (bcspragu) [inline db/sqldb/user.go:70]: As far as I can tell, this function is only used internally and should probably not be exported.

Similar to above comment, we should really have exactly one public function that can create users - it's a somewhat complicated operation and we don't want multiple places doing it differently.

File: db/sqldb/user.go
Comment (bcspragu) [inline db/sqldb/user.go:66]: I'd argue this should happen at the server, not the DB. It's a field, like any other.

File: deps.bzl
Comment (bcspragu) [inline deps.bzl:551]: I know the open source ecosystem runs on herculean feats of individuals, but I'm still not a fan of random people's not-particularly-widely-used libraries being used in PII sensitive contexts (or really just on the backend in general). If you've audited this and trust it, LGTM, but otherwise, I'd fork it or just make the our own minimal version with the normalizations we care about.

I noted that email normalizing is a fraught endeavor, and I'd argue that farming it out to a random dependency is also a fraught solution.

File: frontend/components/standard/Nav.vue
Comment (bcspragu) [inline frontend/components/standard/Nav.vue:18]: Remove?

File: frontend/composables/useSession.ts
Comment (bcspragu) [inline frontend/composables/useSession.ts:33]: Remove?

File: frontend/composables/useSession.ts
Comment (bcspragu) [inline frontend/composables/useSession.ts:66]: As pretty much this whole app should require login, I'd actually just not implement/expose this for now, I've always found this to be kludgy. Do I have an alternative in mind? Definitely not.

File: openapi/pacta.yaml
Comment (bcspragu) [inline openapi/pacta.yaml:330]: `EmptySuccess` Is not what I'd call getting a 401, I'd make a new type (or just leave `content` blank!)

File: pacta/email.go
Comment (bcspragu) [inline pacta/email.go:14]: supernit: `CanonicalizeEmail`?

File: pacta/email.go
Comment (bcspragu) [inline pacta/email.go:21]: Hold up now: https://en.wikipedia.org/wiki/Email_address#Internationalization

Perfectly reasonable email addresses can have non-ASCII characters in them. Again, thinking of our European users, I'd expect umlauts and other things to appear in addresses, which are valid UTF-8 and valid email addresses by [RFC 6532](https://www.rfc-editor.org/rfc/rfc6532).

If internationalization was done via [punycode](https://en.wikipedia.org/wiki/Punycode), like domain names, then the ASCII check would be fine, but it's just full on Unicode

File: pacta/email.go
Comment (bcspragu) [inline pacta/email.go:42]: Remove, though per the previous comment this whole function should prob go

File: pacta/email.go
Comment (bcspragu) [inline pacta/email.go:26]: I feel like this is one of those all downside, no upside checks and we should just do email verification (which I think we already do with Azure). Since [(quoted) spaces are allowed by the email spec](https://en.wikipedia.org/wiki/Email_address#Local-part), not allowing them is just guaranteeing we make _someone's_ life very annoying for no reason.

File: scripts/run_db.sh
Comment (bcspragu) [inline scripts/run_db.sh:106]: :grimacing:

**Review by bcspragu — APPROVED** [review]

File: db/sqldb/user.go
Comment (gbdubs) [reply db/sqldb/user.go:66]: SGTM.

File: deps.bzl
Comment (gbdubs) [reply deps.bzl:551]: I share your general skepticism, which is why I also audited it. RE: should we create our own fork, I was thinking that Bazel deps provides a sufficient mechanism, since the version + checksum validate that this is the only version that we'll ever use unless we explicitly upgrade it. Is there a way we could more firmly lock in at this version?

File: frontend/components/standard/Nav.vue
Comment (gbdubs) [reply frontend/components/standard/Nav.vue:18]: GC

File: frontend/composables/useSession.ts
Comment (gbdubs) [reply frontend/composables/useSession.ts:33]: GC

File: frontend/composables/useSession.ts
Comment (gbdubs) [reply frontend/composables/useSession.ts:66]: I think we're going to still need it - examples include public initiatives, and publicly-shared resources.

File: openapi/pacta.yaml
Comment (gbdubs) [reply openapi/pacta.yaml:330]: Fair - previously I couldn't figure out how the return types work with missing `content`! Feel free to change in your upcoming PR to revolutionize errors!

File: pacta/email.go
Comment (gbdubs) [reply pacta/email.go:14]: TIL! https://wikidiff.com/canonize/canonicalize

Will change.

File: pacta/email.go
Comment (gbdubs) [reply pacta/email.go:21]: Wait I'm confused - that spec seems to claim that it DOES have to be utf-8 encoded, no?

https://en.wikipedia.org/wiki/International_email#:~:text=The%20set%20of,fully%20internationalized%20email.

File: pacta/email.go
Comment (gbdubs) [reply pacta/email.go:42]: GC on removing this line ;)

File: scripts/run_db.sh
Comment (gbdubs) [reply scripts/run_db.sh:106]: Do you have a pointer to the font you use for our other ones? I'd love for them to be consistent.

Comment (bcspragu) [reply scripts/run_db.sh:106]: We link to it! Is called Pagga:

https://github.com/RMI-PACTA/app/blob/2525f7f39e0748935a0f17b1174aa717a469c0da/cmd/server/main.go#L153

File: deps.bzl
Comment (bcspragu) [reply deps.bzl:551]: I think code review is the best/easiest tool we have available, the dependency should only change versions if we `go get -u ...` some other dep that requires a newer version, which seems unlikely.

File: cmd/server/pactasrv/user.go
Comment (bcspragu) [reply cmd/server/pactasrv/user.go:92]: The other reasons are something like:

1. We shouldn't have exported methods that nobody is using, it increases the package's surface area unnecessarily
2. Given the way our login system works ('get token, send to server'), the only way of account creation that makes sense is `GetOrCreateUser[...]`, we shouldn't make it tempting to create accounts any other way

Comment (gbdubs) [reply cmd/server/pactasrv/user.go:92]: Wait I might be misunderstanding. When does the token from Azure get sent directly to the server? Aren't we always facilitating that exchange through the Frontend? If so wouldn't just having getOrCreateUser make sense at the FE layer too?

Agreed on the idea of one method at the DB layer. - I'll make `CreateUser` testonly.

File: db/sqldb/user.go
Comment (gbdubs) [reply db/sqldb/user.go:70]: agreed. fixed.

File: deps.bzl
Comment (gbdubs) [reply deps.bzl:551]: SGTM.

File: openapi/pacta.yaml
Comment (gbdubs) [reply openapi/pacta.yaml:330]: I'll leave this to you :)

File: pacta/email.go
Comment (gbdubs) [reply pacta/email.go:14]: Done.

File: pacta/email.go
Comment (gbdubs) [reply pacta/email.go:21]: VERY GOOD CATCH. Thank you!

File: pacta/email.go
Comment (gbdubs) [reply pacta/email.go:26]: SG

File: scripts/run_db.sh
Comment (gbdubs) [reply scripts/run_db.sh:106]: <3

File: cmd/server/pactasrv/user.go
Comment (bcspragu) [reply cmd/server/pactasrv/user.go]: Copying image from Slack:
image

Comment (bcspragu) [reply cmd/server/pactasrv/user.go]: The token that would go to a `PACTA` `/login` endpoint would be the RMI-specific token from step 4, and that would get or create the user

**Pull request merged**
